### PR TITLE
Use cold functions for panic formatting Option::expect, Result::unwra…

### PR DESCRIPTION
**Pull request by bluss**
Use cold functions for panic formatting Option::expect, Result::unwrap, expect

These methods are marked inline, but insert a big chunk of formatting
code, as well as other error path related code, such as
deallocating a std::io::Error if you have one.

We can explicitly separate out that code path into a function that is
never inline, since the panicking case should always be rare.

### CODE REVIEW

**Comment by bluss** [comment]
Playpen code to compare these in artificial setting: https://play.rust-lang.org/?gist=94e1ab9ed151c75a5a97&version=nightly

`Option::unwrap` is already good. Look at the optimized versions of `option_expect` and `option_expect_new` to compare them. `option_expect_new` now works just like Option::unwrap does, with less code even in the `Some` case too.

**Comment by nagisa** [comment]
Looks pretty good to me. For `Result::expect` this saves ~80 instructions (improvement around the 65% mark) for the optimised assembly in the call location.

**Comment by nikomatsakis** [comment]
Seems like a good idea to me. (Though I still dislike `Option::expect` with the passion of...well, not that much passion really. But a little.)

**Comment by bluss** [comment]
With this change I might even become an Option.expect() convert.

File: src/libcore/option.rs [hunk 1]
```diff
@@ -295,10 +295,16 @@ impl<T> Option<T> {
...
-            None => panic!("{}", msg),
+            None => Self::expect_failed(msg),
         }
     }
 
+    #[inline(never)]
+    #[cold]
+    fn expect_failed(msg: &str) -> ! {
+        panic!("{}", msg)
+    }
```
Comment (alexcrichton) [inline src/libcore/option.rs]: Could this be moved outside of `Option<T>`? That should mean it's not monomorphized and should be even smaller I believe

File: src/libcore/result.rs [hunk 2]
```diff
@@ -684,11 +684,16 @@ impl<T, E: fmt::Debug> Result<T, E> {
...
-                panic!("called `Result::unwrap()` on an `Err` value: {:?}", e)
+            Err(e) => Self::unwrap_failed(e),
         }
     }
 
+    #[inline(never)]
+    #[cold]
+    fn unwrap_failed(error: E) -> ! {
+        panic!("called `Result::unwrap()` on an `Err` value: {:?}", error)
+    }
```
Comment (alexcrichton) [inline src/libcore/result.rs]: One other possible option you could play around with is seeing if this improves codegen by moving it outside the `impl` block and taking `&fmt::Display` to ensure it's just a small function call

Comment (alexcrichton) [inline src/libcore/result.rs]: (and this same strategy I think applies to the bits below as well)

File: src/libcore/option.rs (hunk 1 above)
Comment (bluss) [inline src/libcore/option.rs]: Good catch, didn't think of that. It applies to everything I think.

**Comment by alexcrichton** [comment]
Nice ideas! I had a few more other ideas about monomorphization and seeing if we can't codegen these in libstd, but otherwise I'm always excited about changes like these.

File: src/libcore/result.rs (hunk 2 above)
Comment (bluss) [inline src/libcore/result.rs]: Yes. I've only compared taking `E` vs `&E`. Taking `E` has the benefit that the deallocation code (for String or std::io::Error) moves into the cold path as well..

**Comment by bluss** [comment]
Maybe Result's `unwrap_failed<E>` should move out of the impl block too, is it unnecessarily monomorphized per `T` otherwise?

File: src/libcore/result.rs (hunk 2 above)
Comment (alexcrichton) [inline src/libcore/result.rs]: Oh that's a good point, moving the dtor to the cold path is likely worth it

**Comment by brson** [comment]
Neat find.

**Comment by arthurprs** [comment]
:clap: I like this

File: src/libcore/result.rs [hunk 3]
```diff
@@ -685,7 +685,7 @@ impl<T, E: fmt::Debug> Result<T, E> {
         match self {
             Ok(t) => t,
             Err(e) =>
-                panic!("called `Result::unwrap()` on an `Err` value: {:?}", e)
+                unwrap_failed("called `Result::unwrap()` on an `Err` value: {:?}", e),
```
Comment (alexcrichton) [inline src/libcore/result.rs]: I think the trailing `{:?}` here canb e removed

Comment (alexcrichton) [inline src/libcore/result.rs]: (same below as well

Comment (bluss) [inline src/libcore/result.rs]: It's supposed to be more consistent with a trailing comma for every case. Not sure what rustfmt usually does.

Comment (bluss) [inline src/libcore/result.rs]: oh.. reading.. how do I?

I see. That's an important change.

Comment (alexcrichton) [inline src/libcore/result.rs]: Oh I think that the trailing `: {:?}` part of the string can be removed in both cases, that's covered by the format string at the very bottom

**Comment by bluss** [comment]
I will rebase to squash all the fixups together into one comimt.

**Comment by bluss** [comment]
r? @alexcrichton 

This is done, unless we want to measure the difference of something a bit more interesting. I don't really have the resources to build all of rust multiple times though.

**Comment by hudson-ayers** [comment]
I realize this is a very old PR, but I wanted to point out one downside of this change:

With this change, for `expect()` messages, even if your panic handler does nothing (i.e. `panic_immediate_abort`), the messages passed to `expect()` are still embedded in your binary. 

Consider the two following code snippets:
```rust
fn check(opt: Option<u32>) -> u32 {
    opt.expect("This option should always be Some by the time this function is called.")
}
```
```rust
fn check(opt: Option<u32>) -> u32 {
    match opt {
        Some(val) => return val,
        None => panic!("This option should always be Some by the time this function is called.")
    }
}
```

I would typically assume these two functions should be identical, but in a binary with no panic handler only the former pays the code size cost of the message being embedded in the binary.

Of course, in a binary that has many calls to `expect()` and *does* have an actual panic handler, this change generates substantial savings. But in severely size-constrained binaries with no panic handler, it is a shame that messages passed to `expect()` cannot be optimized out.

**Pull request merged**